From a86f8ea6da4fbef94b7fb77cec1647ed1c471916 Mon Sep 17 00:00:00 2001
From: donaldduck70 <spigot@homemp3.dyndns.org>
Date: Sun, 14 Jan 2018 19:58:58 +0100
Subject: [PATCH] tick entities in sticky chunks


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 740633d9..ceca5760 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1609,7 +1609,7 @@ public abstract class World implements IBlockAccess {
         // TODO: Go back to Vanilla behaviour when comfortable
         // Spigot start
         // Chunk startingChunk = this.getChunkIfLoaded(MathHelper.floor(entity.locX) >> 4, MathHelper.floor(entity.locZ) >> 4);
-        if (flag && !org.spigotmc.ActivationRange.checkIfActive(entity)) {
+        if (flag && !org.spigotmc.ActivationRange.checkIfActive(entity) && !getChunkAt(((int) entity.locX) >> 4, ((int) entity.locZ) >> 4).isSticky()) { // Spigot - if chunk sticky, entities shall be active in it
             entity.ticksLived++;
             entity.inactiveTick();
             // Spigot end
-- 
2.15.1

